test/validation/api/ipsec/ipsec_test_out.c

```diff
@@ -1181,6 +1201,33 @@ static void ipsec_test_capability(void)
 	CU_ASSERT(odp_ipsec_capability(&capa) == 0);
 }
 
+static void test_ipsec_stats(void)
```

Comment:
This is now quite similar to test_esp_out_in_all() except that statistics checking is also enabled. I do not see why we shouldn't just have test_esp_out_in_all() enable statistics checking and be done with it?

Reply:
That is true. May be let me highlight why I intend to do with different function.

1. Currently only success case is covered. I would like to cover other cases also. (proto error, auth error etc), which will look like deviating from what test_esp_out_in_all() is supposed to do. I'll push another version which would cover proto_err, then the change might be more clear to you.
2. We cannot try both success and error case with same run. Multiple runs of one algo is required, but not necessarily all algos as in test_esp_out_in_all()

Reply:
Ok, let me see a version with some error stat test and it may become more clear.

Reply:
I still do not see the benefit of this separate test function. Now the code spends a lot of effort in making it so that statistics are tested only with single algorithm combo when it could just run the test with all algorithms and have better coverage. Why is this better than e.g. having test_esp_out_in() test the statistics?

Reply:
Currently test_esp_out_in() would only try for success cases. So, yes, we can have stats run for every success algo. But then, if we want stats of any error case, then the entire suite need to be re-run with error simulated for each case, right? The effort in making sure only once the test case is run is actually minimal. Majority of the changes proposed is to make sure we don't replicate the code for just this test case.

Currently, we test IPsec features with random algo combinations. Like, few tests use AES-GCM, while few others use NULL+SHA256. What if the implementation doesn't support that combination? I was trying to answer that. Also, I saw that test_esp_out_in() routine was actually an attempt to have a common framework that iterates through all supported ciphers. The intention is to extend the same framework but pick a supported algo combination for all further tests.

Reply:
We are discussing two related points here: 1) whether statistics should be tested with all algorithm combos, 2) whether the statistics test could be combined with the existing test.

For 1) I would say we should test with all algorithms, since we can do it basically without any extra effort. Just picking the first algorithm that happens to be supported may not give good coverage. E.g. combined mode algorithms might behave a bit differently from the traditional ones for some statistics. The current proposal in this patch would actually run the test with null encryption and null authentication which is not good.

Then regarding 2), I think we could do it because of 1). But If you want to keep it as a separate test, that is ok too. In this case the code gets still simpler because because of 1) you would need the separate phase for finding the first supported algorithm (which should anyway have been done using ipsec_check_esp() instead of rerunning the original test).

Reply:
For 1) Yes. You are right. Different algos could be done differently. So running for stats success case for all ciphers do make sense. I'll have this addressed in next version.

For 2) My only problem is the simulation of an error case. Right now, if I add the flags for proto_err, then the success case won't be tested. So only option is to first run the cases where packets would be successfully processed and then proceed to error cases (one by one). Hence my preference to introduce a separate function for stats testing.

Now regarding testing proto_err stats, do you recommend testing it for specific algos (like let's say AES-GCM) or any supported algo or for all algos?

Reply:
> Now regarding testing proto_err stats, do you recommend testing it for specific algos (like let's say AES-GCM) or any supported algo or for all algos?

I would run all stats tests for all algos because it can be easily done. Except maybe not everything (or anything) for null algos (e.g. there cannot be auth error with null authentication, so you cannot test it with that).

Reply:
Agreed. I can try to extend the error cases to run all algos.